transport/src/test/java/io/netty/channel/ChannelCloseTest.java

```diff
@@ -0,0 +1,257 @@
+            .childOption(ChannelOption.AUTO_READ, autoRead)
+            .childHandler(new ChannelInitializer<Channel>() {
+                @Override
+                protected void initChannel(Channel ch) throws Exception {
+                    ChannelPipeline pipeline = ch.pipeline();
+                    pipeline.addLast(handlers);
+                }
+            });
+
+        return serverBootstrap.bind(0)
+                .syncUninterruptibly()
+                .channel();
+    }
+
+    private Channel newClient(Channel server) {
+        Bootstrap bootstrap = new Bootstrap();
+
+        bootstrap.group(GROUP)
+            .channel(NioSocketChannel.class)
+            .option(ChannelOption.SO_REUSEADDR, true)
+            .option(ChannelOption.CONNECT_TIMEOUT_MILLIS, 1000)
+            .handler(new ChannelInboundHandlerAdapter() {
+                @Override
+                public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                    ReferenceCountUtil.release(msg);
+                    throw new IllegalStateException();
+                }
+            });
+
+        return bootstrap.connect(server.localAddress())
+                .syncUninterruptibly()
+                .channel();
+    }
+
+    /**
+     * Scenario 0
+     *
+     * Same as scenario 1 and 2 but we do nothing to disable auto reading
+     */
+    @Test
+    public void testDisconnectDoNothingWithAutoRead() throws Exception {
+        final CountDownLatch latch = new CountDownLatch(1);
+        DisconnectHandler disconnect = new DisconnectHandler();
+
+        ChannelInboundHandlerAdapter handler = new ChannelInboundHandlerAdapter() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) {
+                ReferenceCountUtil.release(msg);
+                latch.countDown();
+            }
+        };
+
+        Channel server = newServer(true, handler, disconnect);
+        Channel client = newClient(server);
+
+        try {
+            // The clients sends the message
+            client.writeAndFlush(newMessage())
+                .syncUninterruptibly();
+
+            // The server receives the message
+            assertTrue(latch.await(1L, TimeUnit.SECONDS));
+
+            // The client disconnects
+            client.close().syncUninterruptibly();
+
+            // The server should get notified
+            if (!disconnect.await(1L, TimeUnit.SECONDS)) {
+                fail("The server didn't receive a disconnect event.");
+            }
+        } finally {
+            client.close();
+            server.close();
+        }
+    }
+
+    /**
+     * Scenario 1
+     *
+     * 1. The server starts with auto reading turned on
+     * 2. The client connects
+     * 3. The client sends a message
+     * 4. The server receives the message
+     * 5. The server turns off auto reading
+     * 6. The client disconnects
+     * 7. The server (should) receive an channelInactive() event
```

Comment:
@Scottmitch would it theoretically be possible for me to miss out on channelInactive events if I set auto read in this manner?

```java
    private void sendToClient(ByteBuf buffer, boolean closeAfterWrite) {
        if (closeAfterWrite) {
            channel.config().setAutoRead(false);
            ChannelFuture future = channel.writeAndFlush(buffer);
            future.addListener(ChannelFutureListener.CLOSE);
        } else {
            channel.writeAndFlush(buffer, channel.voidPromise());
        }
    }
```